File: backend/edugate-core/src/main/java/de/svws_nrw/edugate/core/domain/Schema.java
package de.svws_nrw.edugate.core.domain;

import java.time.Instant;
import java.util.UUID;

/**
 * Ein SVWS-Schema / eine Schuldatenbank (ADR-012): der Datenbestand einer Schule für eine
 * bestimmte {@code umgebung} (z. B. Produktiv, Test, Schulung) auf einer {@code svws_instanz}.
 * Mandantenbezogen über {@code tenantId} (Schulträger der Schule), RLS-relevant. {@code umgebung}
 * ist bewusst freier Text statt eines geschlossenen Enums: Betreiber brauchen erweiterbare Werte
 * über die Startwerte hinaus (nur "PRODUKTIV" bleibt als Konstante fachlich ausgezeichnet, siehe
 * {@link SchemaNamingKonstanten#PRODUKTIV}). {@code status} ist davon getrennt der technische
 * Lebenszyklus ({@link SchemaStatus}).
 */
public record Schema(
    UUID id,
    UUID tenantId,
    UUID schuleId,
    UUID instanzId,
    String schemaName,
    String umgebung,
    SchemaStatus status,
    boolean aktiv,
    String beschreibung,
    SchemaSource source,
    Instant lastSyncedAt,
    Instant createdAt,
    Instant updatedAt
) {
}

File: frontend/src/types/schema.ts
export type SchemaStatus =
  'GEPLANT' | 'VORHANDEN' | 'AKTIV' | 'DEAKTIVIERT' | 'MIGRATION_ERFORDERLICH' | 'FEHLER' | 'ARCHIVIERT'

export type SchemaSource = 'MANUELL' | 'SYNCHRONISIERT'

// `umgebung` ist bewusst freier Text statt eines geschlossenen Enums (ADR-012). Die zulässigen
// Werte werden nicht mehr als Frontend-Konstante gepflegt, sondern über die Betreiber-Einstellung
// "Umgebungen verwalten" (schemaUmgebungStore, Tabelle schema_umgebung). "PRODUKTIV" bleibt dort
// als Systemeintrag der einzige fachlich ausgezeichnete Wert (kein Namenssuffix, höchstens ein
// aktives Schema je Schule).

export interface Schema {
  id: string
  schuleId: string
  instanzId: string
  schemaName: string
  umgebung: string
  status: SchemaStatus
  aktiv: boolean
  beschreibung: string | null
  source: SchemaSource
  lastSyncedAt: string | null
  createdAt: string
  updatedAt: string
}

export interface SchemaCreateFormData {
  instanzId: string
  schemaName: string
  umgebung: string
  beschreibung?: string
}

export interface SchemaUpdateFormData {
  instanzId: string
  schemaName: string
  umgebung: string
  beschreibung?: string
  status: SchemaStatus
  aktiv: boolean
}

export interface SchemaNamingSuggestion {
  schemaName: string | null
  belastbareSchulnummer: boolean
}

/**
 * Ergebnis eines Löschversuchs eines echten Schemas über die SVWS-Privileged-API. Ein
 * fachlicher Fehlschlag (fehlende Zugangsdaten, SVWS-seitige Ablehnung, Netzwerkfehler) ist kein
 * HTTP-Fehler - die Anfrage liefert 200 mit success=false und einer erklärenden message.
 */
export interface SchemaDestroyResult {
  success: boolean
  message: string | null
}
